feat: Add "Chunks" to the Mutable Buffer

alamb opened the pull request:
Related to #579 

Rationale: see https://github.com/influxdata/influxdb_iox/issues/579

Here is the new layout:
```
┌───────────────────────────────────────────────┐
│                                               │
│    ┌────────────────┐                         │
│    │    Database    │                         │
│    └────────────────┘                         │
│             │ one partition per               │
│             │ partition_key                   │
│             ▼                                 │
│    ┌────────────────┐                         │
│    │   Partition    │                         │
│    └────────────────┘                         │
│             │  one mutable Chunk              │
│             │  0 or more immutable            │
│             ▼  Chunks                         │
│    ┌────────────────┐                         │
│    │     Chunk      │                         │
│    └────────────────┘                         │
│             │  multiple Tables (measurements) │
│             ▼                                 │
│    ┌────────────────┐                         │
│    │     Table      │                         │
│    └────────────────┘                         │
│             │  multiple Colums                │
│             ▼                                 │
│    ┌────────────────┐                         │
│    │     Column     │                         │
│    └────────────────┘                         │
│                              MutableBuffer    │
│                                               │
└───────────────────────────────────────────────┘
```

Prior to this PR the code looked like

```
┌───────────────────────────────────────────────┐
│                                               │
│ ┌────────────────┐                            │
│ │    Database    │                            │
│ └────────────────┘                            │
│          │    one partition per               │
│          │    partition_key                   │
│          ▼                                    │
│ ┌────────────────┐                            │
│ │   Partition    │                            │
│ └────────────────┘                            │
│          │      multiple Tables               │
│          │       (measurements)               │
│          ▼                                    │
│ ┌────────────────┐                            │
│ │     Table      │                            │
│ └────────────────┘                            │
│          │      multiple Colums               │
│          ▼                                    │
│ ┌────────────────┐                            │
│ │     Column     │                            │
│ └────────────────┘                            │
│                         MutableBuffer         │
│                                               │
└───────────────────────────────────────────────┘
```

I tried to break this PR into some self contained commits to try and make it easier to follow, but it is still pretty bug
1. Renames Partition --> Chunk
2. Add a new partition layer that has an active/mutable chunk and some number of inactive chunks (wrapped in Arcs)
4. Update the mutable buffer database to keep a map of partition_key --> Partition 
5.  Disconnects the existing local file based WAL from the mutable buffer (the server has been changed to not instantiate a mutable buffer with a WAL in https://github.com/influxdata/influxdb_iox/pull/569, but the code was still present), per @pauldix 
4.  Updates the snapshotting code to use the new notion of Partition.

Note: the dictionary (string interning) is done on a per-chunk basis rather than per-partition -- this is needed so that  as soon as the chunk becomes "immutable" the corresponding dictionary also becomes immutable

#579 calls for some additional APIs to roll the partitions over, etc. Given there doesn't seem to be any tests for snapshotting (which would be the consumer of these APIs), I plan to add them as part of a follow on PR

alamb commented on mutable_buffer/Cargo.toml:
```diff
@@ -11,7 +11,6 @@ data_types = { path = "../data_types" }
 generated_types = { path = "../generated_types" }
 influxdb_line_protocol = { path = "../influxdb_line_protocol" }
 query = { path = "../query" }
-wal = { path = "../wal" }
```
Here is removing the use of the file based WAL from the mutable_buffer

alamb commented on mutable_buffer/Cargo.toml:
```diff
@@ -26,7 +25,3 @@ tracing = "0.1"
 criterion = "0.3"
-
-[[bench]]
-name = "benchmark"
```
The only benchmarks in this crate measure the speed of recovery from the WAL

alamb commented on mutable_buffer/src/chunk.rs:
```diff
@@ -0,0 +1,475 @@
+//! Represents a Chunk of data (a collection of tables and their data within
```
Github didn't realize it, but this file is mostly what is in https://github.com/influxdata/influxdb_iox/blob/main/mutable_buffer/src/partition.rs (as that mostly corresponded to a chunk)

alamb commented on mutable_buffer/src/database.rs:
```diff
@@ -1398,88 +1274,95 @@ mod tests {
         Ok(())
     }
 
-    #[tokio::test]
```
I figured these tests might be cool to repurpose when testing recovery from WAL segments -- but I can just delete them entirely if reviewers feel that would be better

pauldix replied:
I think the structure of recovery is going to change enough that the tests will likely have to be rewritten. I also have on my todo list (or for someone else) to do a little bit of benchmarking of Flatbuffers vs. JSON for the WAL format, which will likely test this recovery path as well. So probably fine to delete for now.

alamb commented on mutable_buffer/src/lib.rs:
```diff
@@ -1,4 +1,7 @@
-//! Contains an in memory write buffer that stores incoming data, durably.
```
Github is making this file ~ impossible to read -- but if you look at it locally, it simply has a `mutable_chunk` and a list of `immutable_chunks` and a method to roll them over (not yet implemented) which I will implement in the next PR

pauldix commented on mutable_buffer/src/chunk.rs:
```diff
@@ -0,0 +1,475 @@
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Chunk {} not found", chunk))]
+    ChunkNotFound { chunk: String },
```
Looks like this error can be removed since this was looking up a partition by its key. And it looks like it's never used.

pauldix commented on mutable_buffer/src/lib.rs:
```diff
@@ -1,4 +1,7 @@
+//! Contains an in memory mutable buffer that stores incoming data in
+//! a structure that is designed to be quickly appended to as well as queried
+//!
+//! The mutable buffer is
```
Looked at it locally and I'm seeing this. Looks like something got lost? The mutable buffer is... what? Please tell me?! I MUST KNOW!

alamb replied:
hahaha -- whoops -- yes I will complete that thought.

pauldix commented on mutable_buffer/src/chunk.rs:
```diff
@@ -0,0 +1,475 @@
+#[derive(Debug)]
+pub struct Chunk {
+    /// Chunk key for all rows in this the chunk
+    pub key: String,
```
I wouldn't expect Chunk to have a key (which I see is initialized with the partition key). You could have `partition_key` in here, but Chunks should definitely have an ID `u64`. The chunk id should be a monotonically increasing number within the scope of the partition. What I'm anticipating is that this is part of the path in object storage. e.g. `<writer id>/data/<database>/<partition key>/<chunk id>/<parquet files>`

pauldix commented on mutable_buffer/src/chunk.rs:
```diff
@@ -0,0 +1,475 @@
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Debug)]
+pub struct Chunk {
```
I think chunk should have a created at time, which should be the time at which it receives its first write. Also should have a rollover time, but it looks like you're planning on doing that in a follow on PR.

alamb replied:
Yes I thought that would be good.

pauldix reviewed (approved):
Left a few comments. Dictionary per chunk is great. Overall looks really good and the ASCII art in the PR description is helpful. Might be good to put that into the doc comments for the mutable buffer crate.

Since this touches so much, I'd like to shoot to get this merged as quickly as possible and then fixup stuff in follow on PRs. Maybe a few touchups based on my comments and then :shipit:!

alamb commented:
Thanks @pauldix  -- I plan to address these comments first thing tomorrow and get this merged, and then get the follow on PRs up.

alamb commented:
I filed https://github.com/influxdata/influxdb_iox/issues/598 to track adding ID's to the chunks

alamb commented:
Ok, I have cleaned up the comments, removed unused error types and deleted the old wal tests. I plan to merge this once the CI is done and add the appropriate Ids / rollover logic in a subsequent PR
